// @colony/colony-js/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Versions of colonyNetwork
var ColonyVersion;
(function (ColonyVersion) {
    ColonyVersion[ColonyVersion["GoerliGlider"] = 1] = "GoerliGlider";
    ColonyVersion[ColonyVersion["Glider"] = 2] = "Glider";
    ColonyVersion[ColonyVersion["AuburnGlider"] = 3] = "AuburnGlider";
    ColonyVersion[ColonyVersion["BurgundyGlider"] = 4] = "BurgundyGlider";
})(ColonyVersion = exports.ColonyVersion || (exports.ColonyVersion = {}));
exports.CurrentVersion = ColonyVersion.BurgundyGlider;
var Network;
(function (Network) {
    Network["Mainnet"] = "mainnet";
    Network["Goerli"] = "goerli";
    Network["Local"] = "local";
})(Network = exports.Network || (exports.Network = {}));
var ColonyRole;
(function (ColonyRole) {
    ColonyRole[ColonyRole["Recovery"] = 0] = "Recovery";
    ColonyRole[ColonyRole["Root"] = 1] = "Root";
    ColonyRole[ColonyRole["Arbitration"] = 2] = "Arbitration";
    ColonyRole[ColonyRole["Architecture"] = 3] = "Architecture";
    // eslint-disable-next-line @typescript-eslint/camelcase
    ColonyRole[ColonyRole["ArchitectureSubdomain_DEPRECATED"] = 4] = "ArchitectureSubdomain_DEPRECATED";
    ColonyRole[ColonyRole["Funding"] = 5] = "Funding";
    ColonyRole[ColonyRole["Administration"] = 6] = "Administration";
})(ColonyRole = exports.ColonyRole || (exports.ColonyRole = {}));
var ClientType;
(function (ClientType) {
    ClientType["ColonyClient"] = "ColonyClient";
    ClientType["NetworkClient"] = "NetworkClient";
    ClientType["OneTxPaymentClient"] = "OneTxPaymentClient";
    ClientType["OneTxPaymentFactoryClient"] = "OneTxPaymentFactoryClient";
    ClientType["TokenClient"] = "TokenClient";
    ClientType["TokenLockingClient"] = "TokenLockingClient";
})(ClientType = exports.ClientType || (exports.ClientType = {}));
var TokenClientType;
(function (TokenClientType) {
    TokenClientType["Colony"] = "Colony";
    TokenClientType["Erc20"] = "Erc20";
    TokenClientType["Dai"] = "Dai";
})(TokenClientType = exports.TokenClientType || (exports.TokenClientType = {}));
exports.colonyNetworkAddresses = {
    [Network.Mainnet]: '0x5346D0f80e2816FaD329F2c140c870ffc3c3E2Ef',
    [Network.Goerli]: '0x79073fc2117dD054FCEdaCad1E7018C9CbE3ec0B',
    [Network.Local]: undefined,
};
exports.oneTxPaymentFactoryAddresses = {
    [Network.Mainnet]: '0x6fb63009e3e03cbf6917647d64ad81939f267067',
    [Network.Goerli]: '0xe4aaf13b6a3d84d9d025ef1e8aa1873ecb027322',
    [Network.Local]: undefined,
};
exports.ROOT_DOMAIN_ID = 1;
var FundingPotAssociatedType;
(function (FundingPotAssociatedType) {
    FundingPotAssociatedType[FundingPotAssociatedType["Unassigned"] = 0] = "Unassigned";
    FundingPotAssociatedType[FundingPotAssociatedType["Domain"] = 1] = "Domain";
    FundingPotAssociatedType[FundingPotAssociatedType["Task"] = 2] = "Task";
    FundingPotAssociatedType[FundingPotAssociatedType["Payment"] = 3] = "Payment";
    FundingPotAssociatedType[FundingPotAssociatedType["Expenditure"] = 4] = "Expenditure";
})(FundingPotAssociatedType = exports.FundingPotAssociatedType || (exports.FundingPotAssociatedType = {}));
exports.REPUTATION_ORACLE_ENDPOINT = 'https://colony.io/reputation';
exports.tokenAddresses = {
    SAI: '0x89d24A6b4CcB1B6fAA2625fE562bDD9a23260359',
};
